types/wordpress__blocks/index.d.ts

```diff
@@ -70,6 +76,32 @@ export interface Block<T extends Record<string, any> = {}> {
      * Component to render in the editor.
      */
     readonly edit?: ComponentType<BlockEditProps<T>> | undefined;
+    /**
+     * Block type editor script definition.
+     * It will only be enqueued in the context of the editor.
+     *
+     * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#editor-script}
+     */
+    readonly editorScript?: string;
+    /**
+     * Block type editor style definition.
+     * It will only be enqueued in the context of the editor.
+     *
+     * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#editor-style}
+     */
+    readonly editorStyle?: string;
+    /**
+     * It provides structured example data for the block.
+     *
+     * @see {@link https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#example}
+     */
+    readonly example?: {
+        readonly attributes: {
+            readonly [k in keyof T]: any;
+        };
+        readonly innerBlocks?: any[];
```

Comment:
this isn't really `any[]` here since the `innerBlocks` need to be blocks. does `Block<any>` or `Block<unknown>` work?

Reply:
`Block<any>` would not work since it only needs takes the `name` and `attributes` properties:

```js
example: {
    attributes: {
        cover: 'https://example.com/image.jpg',
    },
    innerBlocks: [
        {
            name: 'core/paragraph',
            attributes: {
                /* translators: example text. */
                content: __(
                    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent et eros eu felis.'
                ),
            },
        },
    ],
},
```
Example from [here](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/#example-optional).

---

Would go with
```ts
readonly innerBlocks?: ReadonlyArray<{ name: string; attributes: Record<string, any> }>;
```
for now.

Reply:
if we only need those properties that seems like we could use this to do that…

```ts
type InnerBlock = Partial<Block> & Pick<Block, 'name' | 'attributes'>;
```

this prevents us from needing to duplicate the definitions of the block types and also preserves the fact that in many (all?) cases those other properties will be there, even if not mandatory for some reason. otherwise we're saying, among other things, that it's not possible to have inner blocks which themselves contain inner blocks.

Reply:
Would generally agree, but the `attributes` type seems different:
While `Block.attributes` expects a declarative object in the form of:
```js
attributes: {
  foo: {
      source: 'text';
      type: 'string';
  }
}
```

The attributes property of the `innerBlocks` seem to be passed by value:
```js
attributes: {
  foo: 'bar'
}
```

Reply:
> While Block.attributes expects a declarative object in the form of:

This is a good point. I think the block attributes types are a big confused at this point. There's the `BlockConfiguration` which takes those attribute declarations, but the `Block` objects themselves take attributes that are the real types, which _cannot_ be those same types as given for the `BlockConfiguration`

don't let this hold up the PR. if we can't reuse `Block['attributes']` as we ideally would, we can fix that later.

Reply:
Ah okay, yes your proposed type seem to work fine with the example.
However looking at the implementation it seems that `innerBlocks` can be recursive:

https://github.com/WordPress/gutenberg/blob/87d761eb48820936f71568eb81d7130be4dfc25a/packages/blocks/src/api/factory.js#L579-L587

I've now updated the type to:
```ts
type BlockExampleInnerBlock = ReadonlyArray<
    Partial<Block> &
        Pick<Block, 'name' | 'attributes'> & {
            innerBlocks?: BlockExampleInnerBlock;
        }
>;
```

Reply:
> it seems that innerBlocks can be recursive:

yep! that's one of the reasons I brought it up.

Reply:
though…we shouldn't limit it to `BlockExampleInnerBlock`. I think in this case `Pick<Block, 'name' | 'attributes' | 'innerBlocks'>` would work, otherwise what we're saying is not just that the block can be recursive, but that it can only be recursive on its own type.

consider the group block; it can have child blocks of any type because its role is to be a kind of neutral container. if we kept this definition then we wouldn't be able to show any meaningful example for a column block because it would demand we only allow `group` blocks to be the children of a group block.

Reply:
my last comment was wrong; sorry for saying otherwise. I didn't see at first what you were doing here.

concretely though I think we might want to drop `attributes` because not every block requires it and I don't want to overconstrain our types on the example code. on the `example` property that means maybe mark it also as optional or do this

```ts
example?: Readonly<Partial<Block> & Pick<Block, 'name'> & {innerBlocks: ReadonlyArray<BlockExampleInnerBlock>}>;
```

that's the only thing I think that would likely need to change here. I don't want to hold up this PR and I'm afraid I made comments pushing you to something that I'm now suggesting walking back from. if you don't agree with me I'll still give my approval - these make the types better. primarily I just want to avoid falsely indicating a type error where none exists.

Reply:
Sorry for the late response, thanks for the detailed explanations 👍
`example.name` path doesn't seem to be required from the [docs](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/#example-optional), so I guess
```ts
example?: Readonly<Partial<Block> & { innerBlocks: ReadonlyArray<BlockExampleInnerBlock> }>;
```
is fine here?